File: python/sorting/perf_measurer.py
import random
import time

from .sort import Sort


class PerfMeasurer:
    def __init__(self, sorter: Sort) -> None:
        self.sorter = sorter

    def reportPerf(self) -> str:
        sizes = [100, 200, 500, 1000, 2000, 5000]

        report = ""

        for size in sizes:
            ls = list(range(size))
            random.seed(0)
            random.shuffle(ls)
            start = time.time()
            self.sorter.sort(ls)
            end = time.time()

            diff = end - start

            report += "Time to sort an array of size {:5d} is {:.10f} sec\n".format(
                size, diff
            )

        return report

File: scala/src/main/scala/sorting/PerfMeasurer.scala
package sorting

import scala.util.Random

class PerfMeasurer(val sorter: Sort) {
  def reportPerf(): String = {

    val builder = new StringBuilder
    val sizes = Array(100, 200, 500, 1000, 2000, 5000, 10000, 20000, 50000)

    for (size <- sizes) {
      var arr: Array[Int] = Array.range(0, size)
      val shuffled = Random.shuffle(arr)

      val start = System.nanoTime()
      sorter.sort(arr)
      val end = System.nanoTime()
      val diff = (end - start).toDouble / 1000 / 1000 / 1000

      builder.append(
        f"Time to sort an array of size $size%5d is $diff%.8f sec\n"
      )
    }

    builder.result()
  }
}
